File: bindings/java/src/main/java/tw/com/fugle/marketdata/FugleWebSocketClient.java
package tw.com.fugle.marketdata;

import tw.com.fugle.marketdata.generated.*;

import java.util.concurrent.*;

/**
 * Idiomatic Java wrapper for WebSocket client with dual streaming patterns.
 *
 * <p>Supports two modes of operation:
 * <ul>
 *   <li><b>Callback mode:</b> Provide a WebSocketListener for push-based events</li>
 *   <li><b>Pull mode:</b> Use poll()/take() to consume messages from a BlockingQueue</li>
 * </ul>
 *
 * <h3>Example - Callback Mode:</h3>
 * <pre>{@code
 * WebSocketListener listener = new WebSocketListener() {
 *     public void onConnected() {
 *         System.out.println("Connected!");
 *     }
 *     public void onDisconnected() {
 *         System.out.println("Disconnected");
 *     }
 *     public void onMessage(StreamMessage message) {
 *         System.out.println("Event: " + message.event());
 *     }
 *     public void onError(String error) {
 *         System.err.println("Error: " + error);
 *     }
 * };
 *
 * try (FugleWebSocketClient client = FugleWebSocketClient.builder()
 *         .apiKey("YOUR_API_KEY")
 *         .stock()
 *         .listener(listener)
 *         .build()) {
 *
 *     client.connect().get();
 *     client.subscribe("trades", "2330").get();
 *     Thread.sleep(10000);
 * }
 * }</pre>
 *
 * <h3>Example - Pull Mode:</h3>
 * <pre>{@code
 * try (FugleWebSocketClient client = FugleWebSocketClient.builder()
 *         .apiKey("YOUR_API_KEY")
 *         .stock()
 *         .queueCapacity(1000)
 *         .build()) {
 *
 *     client.connect().get();
 *     client.subscribe("trades", "2330").get();
 *
 *     while (true) {
 *         StreamMessage msg = client.poll(1, TimeUnit.SECONDS);
 *         if (msg != null) {
 *             System.out.println("Event: " + msg.event());
 *         }
 *
 *         // Check for async errors
 *         if (client.hasErrors()) {
 *             String error = client.pollError();
 *             System.err.println("Async error: " + error);
 *         }
 *     }
 * }
 * }</pre>
 */
public class FugleWebSocketClient implements AutoCloseable {

    private final WebSocketClient webSocketClient;
    private final BlockingQueue<StreamMessage> messageQueue;
    private final BlockingQueue<String> errorQueue;

    private FugleWebSocketClient(WebSocketClient webSocketClient,
                                  BlockingQueue<StreamMessage> messageQueue,
                                  BlockingQueue<String> errorQueue) {
        this.webSocketClient = webSocketClient;
        this.messageQueue = messageQueue;
        this.errorQueue = errorQueue;
    }

    /**
     * Connect to the WebSocket server.
     *
     * @return CompletableFuture that completes when connected
     * @throws ApiException if connection fails
     * @throws AuthException if authentication fails
     */
    public CompletableFuture<Void> connect() {
        return webSocketClient.connect()
                .exceptionally(e -> { throw FugleException.unwrap(e); });
    }

    /**
     * Disconnect from the WebSocket server.
     *
     * @return CompletableFuture that completes when disconnected
     */
    public CompletableFuture<Void> disconnect() {
        return webSocketClient.disconnect();
    }

    /**
     * Check if currently connected.
     */
    public boolean isConnected() {
        try {
            return webSocketClient.isConnected();
        } catch (Exception e) {
            throw FugleException.unwrap(e);
        }
    }

    /**
     * Check if the client has been shut down.
     */
    public boolean isClosed() {
        try {
            return webSocketClient.isClosed();
        } catch (Exception e) {
            throw FugleException.unwrap(e);
        }
    }

    /**
     * Send a ping message to the server.
     *
     * @param state Optional state string echoed back in the pong response (nullable)
     * @return CompletableFuture that completes when the ping is sent
     */
    public CompletableFuture<Void> ping(String state) {
        return webSocketClient.ping(state)
                .exceptionally(e -> { throw FugleException.unwrap(e); });
    }

    /**
     * Query the server for current subscriptions.
     * The response arrives via the message callback or pull queue.
     *
     * @return CompletableFuture that completes when the query is sent
     */
    public CompletableFuture<Void> querySubscriptions() {
        return webSocketClient.querySubscriptions()
                .exceptionally(e -> { throw FugleException.unwrap(e); });
    }

    /**
     * Subscribe to a channel for a symbol.
     *
     * @param channel Channel name (e.g., "trades", "candles", "books")
     * @param symbol Symbol to subscribe (e.g., "2330")
     * @return CompletableFuture that completes when subscribed
     * @throws ApiException if subscription fails
     */
    public CompletableFuture<Void> subscribe(String channel, String symbol) {
        return webSocketClient.subscribe(channel, symbol)
                .exceptionally(e -> { throw FugleException.unwrap(e); });
    }

    /**
     * Unsubscribe from a channel for a symbol.
     *
     * @param channel Channel name
     * @param symbol Symbol to unsubscribe
     * @return CompletableFuture that completes when unsubscribed
     */
    public CompletableFuture<Void> unsubscribe(String channel, String symbol) {
        return webSocketClient.unsubscribe(channel, symbol)
                .exceptionally(e -> { throw FugleException.unwrap(e); });
    }

    /**
     * Poll for the next message (non-blocking).
     *
     * <p>Only available in pull mode (when listener was not provided).
     *
     * @return Next message, or null if queue is empty
     */
    public StreamMessage poll() {
        if (messageQueue == null) {
            throw new IllegalStateException("poll() only available in pull mode (no listener provided)");
        }
        return messageQueue.poll();
    }

    /**
     * Poll for the next message with timeout.
     *
     * <p>Only available in pull mode (when listener was not provided).
     *
     * @param timeout Maximum time to wait
     * @param unit Time unit for timeout
     * @return Next message, or null if timeout expires
     * @throws InterruptedException if interrupted while waiting
     */
    public StreamMessage poll(long timeout, TimeUnit unit) throws InterruptedException {
        if (messageQueue == null) {
            throw new IllegalStateException("poll() only available in pull mode (no listener provided)");
        }
        return messageQueue.poll(timeout, unit);
    }

    /**
     * Take the next message (blocking).
     *
     * <p>Only available in pull mode (when listener was not provided).
     *
     * @return Next message (blocks until available)
     * @throws InterruptedException if interrupted while waiting
     */
    public StreamMessage take() throws InterruptedException {
        if (messageQueue == null) {
            throw new IllegalStateException("take() only available in pull mode (no listener provided)");
        }
        return messageQueue.take();
    }

    /**
     * Get the current message queue size.
     *
     * <p>Only available in pull mode (when listener was not provided).
     *
     * @return Number of messages in queue
     */
    public int queueSize() {
        if (messageQueue == null) {
            throw new IllegalStateException("queueSize() only available in pull mode (no listener provided)");
        }
        return messageQueue.size();
    }

    /**
     * Check if any async errors have occurred.
     *
     * <p>Only available in pull mode (when listener was not provided).
     *
     * @return true if errors are in the error queue
     */
    public boolean hasErrors() {
        if (errorQueue == null) {
            throw new IllegalStateException("hasErrors() only available in pull mode (no listener provided)");
        }
        return !errorQueue.isEmpty();
    }

    /**
     * Poll for the next error (non-blocking).
     *
     * <p>Only available in pull mode (when listener was not provided).
     *
     * @return Next error message, or null if no errors
     */
    public String pollError() {
        if (errorQueue == null) {
            throw new IllegalStateException("pollError() only available in pull mode (no listener provided)");
        }
        return errorQueue.poll();
    }

    @Override
    public void close() {
        webSocketClient.close();
    }

    /**
     * Create a new builder.
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Builder for FugleWebSocketClient.
     */
    public static class Builder {
        private String apiKey;
        private String bearerToken;
        private String sdkToken;
        private String baseUrl;
        private WebSocketEndpoint endpoint = WebSocketEndpoint.STOCK;
        private WebSocketListener listener;
        private int queueCapacity = 10000;
        private ReconnectOptions reconnectOptions;
        private HealthCheckOptions healthCheckOptions;

        private Builder() {}

        /**
         * Set API key for authentication.
         */
        public Builder apiKey(String apiKey) {
            this.apiKey = apiKey;
            return this;
        }

        /**
         * Set bearer token for OAuth authentication.
         */
        public Builder bearerToken(String bearerToken) {
            this.bearerToken = bearerToken;
            return this;
        }

        /**
         * Set SDK token for legacy authentication.
         */
        public Builder sdkToken(String sdkToken) {
            this.sdkToken = sdkToken;
            return this;
        }

        /**
         * Set custom base URL for WebSocket endpoint.
         *
         * @param baseUrl Custom base URL (e.g., "wss://custom.ws.fugle.tw")
         * @return This builder for chaining
         */
        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        /**
         * Use stock market data endpoint (default).
         */
        public Builder stock() {
            this.endpoint = WebSocketEndpoint.STOCK;
            return this;
        }

        /**
         * Use futures and options market data endpoint.
         */
        public Builder futopt() {
            this.endpoint = WebSocketEndpoint.FUT_OPT;
            return this;
        }

        /**
         * Use futures and options market data endpoint (alias).
         */
        public Builder futOpt() {
            return futopt();
        }

        /**
         * Provide a listener for callback mode (push-based events).
         *
         * <p>If a listener is provided, poll()/take() methods will throw IllegalStateException.
         */
        public Builder listener(WebSocketListener listener) {
            this.listener = listener;
            return this;
        }

        /**
         * Set message queue capacity for pull mode.
         *
         * <p>Default: 10000
         * <p>Only used when no listener is provided.
         */
        public Builder queueCapacity(int capacity) {
            this.queueCapacity = capacity;
            return this;
        }

        /**
         * Set reconnection options for WebSocket client.
         *
         * @param reconnectOptions Reconnection configuration
         * @return This builder for chaining
         */
        public Builder reconnect(ReconnectOptions reconnectOptions) {
            this.reconnectOptions = reconnectOptions;
            return this;
        }

        /**
         * Set health check options for WebSocket client.
         *
         * @param healthCheckOptions Health check configuration
         * @return This builder for chaining
         */
        public Builder healthCheck(HealthCheckOptions healthCheckOptions) {
            this.healthCheckOptions = healthCheckOptions;
            return this;
        }

        /**
         * Build the FugleWebSocketClient.
         *
         * @throws FugleException if exactly one authentication method is not provided
         */
        public FugleWebSocketClient build() {
            // Exactly-one-auth validation
            int authCount = 0;
            if (apiKey != null) authCount++;
            if (bearerToken != null) authCount++;
            if (sdkToken != null) authCount++;

            if (authCount == 0) {
                throw new FugleException("Provide exactly one of: apiKey, bearerToken, sdkToken");
            }
            if (authCount > 1) {
                throw new FugleException("Provide exactly one of: apiKey, bearerToken, sdkToken");
            }

            WebSocketListener effectiveListener;
            BlockingQueue<StreamMessage> messageQueue;
            BlockingQueue<String> errorQueue;

            if (listener != null) {
                // Callback mode: use provided listener directly
                effectiveListener = listener;
                messageQueue = null;
                errorQueue = null;
            } else {
                // Pull mode: create internal listener with BlockingQueue
                messageQueue = new LinkedBlockingQueue<>(queueCapacity);
                errorQueue = new LinkedBlockingQueue<>();

                effectiveListener = new InternalListener(messageQueue, errorQueue);
            }

            // TODO: Current UniFFI WebSocketClient constructors only accept api_key
            // For bearerToken/sdkToken support, store values for future use
            // (same pattern as Python/Node.js phases 12-02, 13-02)
            if (apiKey != null) {
                // Convert config options to UniFFI record types
                ReconnectConfigRecord reconnectRecord = null;
                if (reconnectOptions != null) {
                    reconnectRecord = new ReconnectConfigRecord(
                        reconnectOptions.getMaxAttempts() != null ? reconnectOptions.getMaxAttempts() : 0,
                        reconnectOptions.getInitialDelayMs() != null ? reconnectOptions.getInitialDelayMs() : 0L,
                        reconnectOptions.getMaxDelayMs() != null ? reconnectOptions.getMaxDelayMs() : 0L
                    );
                }

                HealthCheckConfigRecord healthCheckRecord = null;
                if (healthCheckOptions != null) {
                    healthCheckRecord = new HealthCheckConfigRecord(
                        healthCheckOptions.getEnabled() != null ? healthCheckOptions.getEnabled() : false,
                        healthCheckOptions.getIntervalMs() != null ? healthCheckOptions.getIntervalMs() : 0L,
                        healthCheckOptions.getMaxMissedPongs() != null ? healthCheckOptions.getMaxMissedPongs() : 0L
                    );
                }

                WebSocketClient client;
                if (baseUrl != null) {
                    client = WebSocketClient.newWithUrl(
                        apiKey, effectiveListener, endpoint, baseUrl, reconnectRecord, healthCheckRecord
                    );
                } else {
                    client = WebSocketClient.newWithConfig(
                        apiKey, effectiveListener, endpoint, reconnectRecord, healthCheckRecord
                    );
                }

                return new FugleWebSocketClient(client, messageQueue, errorQueue);
            } else {
                // bearerToken or sdkToken provided but not yet supported by UniFFI WebSocketClient
                throw new FugleException("WebSocket currently only supports apiKey authentication. " +
                                       "bearerToken and sdkToken support coming in future release.");
            }
        }
    }

    /**
     * Internal listener for pull mode that forwards events to BlockingQueues.
     */
    private static class InternalListener implements WebSocketListener {
        private final BlockingQueue<StreamMessage> messageQueue;
        private final BlockingQueue<String> errorQueue;

        InternalListener(BlockingQueue<StreamMessage> messageQueue,
                        BlockingQueue<String> errorQueue) {
            this.messageQueue = messageQueue;
            this.errorQueue = errorQueue;
        }

        @Override
        public void onConnected() {
            // No action needed in pull mode
        }

        @Override
        public void onDisconnected() {
            // No action needed in pull mode
        }

        @Override
        public void onMessage(StreamMessage message) {
            // Offer to queue (non-blocking, drops if full)
            messageQueue.offer(message);
        }

        @Override
        public void onError(String errorMessage) {
            // Offer to error queue
            errorQueue.offer(errorMessage);
        }

        @Override
        public void onReconnecting(Integer attempt) {
            // No action needed in pull mode
        }

        @Override
        public void onReconnectFailed(Integer attempts) {
            errorQueue.offer("All " + attempts + " reconnection attempts exhausted");
        }
    }
}

File: bindings/csharp/MarketdataUniffi/FugleWebSocketClient.cs
// Public wrapper providing FubonNeo-compatible WebSocket API over UniFFI-generated bindings
using System;
using System.Threading.Tasks;

namespace FugleMarketData
{
    /// <summary>
    /// WebSocket endpoint types for market data streaming.
    /// </summary>
    public enum WebSocketEndpoint
    {
        /// <summary>
        /// Stock market data stream
        /// </summary>
        Stock,

        /// <summary>
        /// Futures and options market data stream
        /// </summary>
        FutOpt
    }

    /// <summary>
    /// Interface for receiving WebSocket events.
    /// Implement this interface to handle streaming market data.
    /// </summary>
    public interface IWebSocketListener
    {
        /// <summary>
        /// Called when WebSocket connection is established.
        /// </summary>
        void OnConnected();

        /// <summary>
        /// Called when WebSocket connection is closed.
        /// </summary>
        void OnDisconnected();

        /// <summary>
        /// Called when a market data message is received.
        /// </summary>
        /// <param name="message">Streaming message with event, channel, symbol, and data</param>
        void OnMessage(uniffi.marketdata_uniffi.StreamMessage message);

        /// <summary>
        /// Called when an error occurs.
        /// </summary>
        /// <param name="errorMessage">Error description</param>
        void OnError(string errorMessage);

        /// <summary>
        /// Called when a reconnection attempt starts.
        /// </summary>
        /// <param name="attempt">Current attempt number (1-based)</param>
        void OnReconnecting(uint attempt);

        /// <summary>
        /// Called when all reconnection attempts are exhausted.
        /// </summary>
        /// <param name="attempts">Total number of attempts made</param>
        void OnReconnectFailed(uint attempts);
    }

    /// <summary>
    /// Internal adapter to convert IWebSocketListener to UniFFI WebSocketListener interface.
    /// </summary>
    internal class WebSocketListenerAdapter : uniffi.marketdata_uniffi.WebSocketListener
    {
        private readonly IWebSocketListener _listener;

        public WebSocketListenerAdapter(IWebSocketListener listener)
        {
            _listener = listener ?? throw new ArgumentNullException(nameof(listener));
        }

        public void OnConnected() => _listener.OnConnected();
        public void OnDisconnected() => _listener.OnDisconnected();
        public void OnMessage(uniffi.marketdata_uniffi.StreamMessage message) => _listener.OnMessage(message);
        public void OnError(string errorMessage) => _listener.OnError(errorMessage);
        public void OnReconnecting(uint attempt) => _listener.OnReconnecting(attempt);
        public void OnReconnectFailed(uint attempts) => _listener.OnReconnectFailed(attempts);
    }

    /// <summary>
    /// WebSocket client for streaming market data.
    /// Provides real-time data via callback-based interface.
    /// </summary>
    /// <example>
    /// <code>
    /// class MyListener : IWebSocketListener
    /// {
    ///     public void OnConnected() => Console.WriteLine("Connected!");
    ///     public void OnDisconnected() => Console.WriteLine("Disconnected");
    ///     public void OnMessage(StreamMessage msg) => Console.WriteLine($"{msg.Channel}: {msg.Symbol}");
    ///     public void OnError(string error) => Console.WriteLine($"Error: {error}");
    /// }
    ///
    /// var listener = new MyListener();
    /// using var client = new WebSocketClient("your-api-key", listener);
    /// await client.ConnectAsync();
    /// await client.SubscribeAsync("trades", "2330");
    /// // Messages will arrive via OnMessage callback
    /// </code>
    /// </example>
    public sealed class WebSocketClient : IDisposable
    {
        private readonly uniffi.marketdata_uniffi.WebSocketClient _inner;
        private bool _disposed;
        private readonly ReconnectOptions? _reconnectOptions;
        private readonly HealthCheckOptions? _healthCheckOptions;

        /// <summary>
        /// Create a WebSocket client for stock market data streaming.
        /// </summary>
        /// <param name="apiKey">Fugle API key</param>
        /// <param name="listener">Listener to receive WebSocket events</param>
        /// <exception cref="ArgumentNullException">If apiKey or listener is null</exception>
        public WebSocketClient(string apiKey, IWebSocketListener listener)
        {
            if (string.IsNullOrEmpty(apiKey))
                throw new ArgumentNullException(nameof(apiKey));
            if (listener == null)
                throw new ArgumentNullException(nameof(listener));

            var adapter = new WebSocketListenerAdapter(listener);
            _inner = new uniffi.marketdata_uniffi.WebSocketClient(apiKey, adapter);
            _reconnectOptions = null;
            _healthCheckOptions = null;
        }

        /// <summary>
        /// Create a WebSocket client for a specific endpoint (stock or futopt).
        /// </summary>
        /// <param name="apiKey">Fugle API key</param>
        /// <param name="listener">Listener to receive WebSocket events</param>
        /// <param name="endpoint">Endpoint type: Stock or FutOpt</param>
        /// <exception cref="ArgumentNullException">If apiKey or listener is null</exception>
        public WebSocketClient(string apiKey, IWebSocketListener listener, WebSocketEndpoint endpoint)
        {
            if (string.IsNullOrEmpty(apiKey))
                throw new ArgumentNullException(nameof(apiKey));
            if (listener == null)
                throw new ArgumentNullException(nameof(listener));

            var adapter = new WebSocketListenerAdapter(listener);
            var uniffiEndpoint = endpoint switch
            {
                WebSocketEndpoint.Stock => uniffi.marketdata_uniffi.WebSocketEndpoint.Stock,
                WebSocketEndpoint.FutOpt => uniffi.marketdata_uniffi.WebSocketEndpoint.FutOpt,
                _ => throw new ArgumentOutOfRangeException(nameof(endpoint))
            };

            _inner = uniffi.marketdata_uniffi.WebSocketClient.NewWithEndpoint(apiKey, adapter, uniffiEndpoint);
            _reconnectOptions = null;
            _healthCheckOptions = null;
        }

        /// <summary>
        /// Create a WebSocket client with configuration options.
        /// Exactly one authentication method must be provided in the options.
        /// </summary>
        /// <param name="options">Configuration options including authentication and connection settings</param>
        /// <param name="listener">Listener to receive WebSocket events</param>
        /// <exception cref="ArgumentNullException">If options or listener is null</exception>
        /// <exception cref="ArgumentException">If zero or multiple authentication methods are provided</exception>
        public WebSocketClient(WebSocketClientOptions options, IWebSocketListener listener)
        {
            if (options == null)
                throw new ArgumentNullException(nameof(options));
            if (listener == null)
                throw new ArgumentNullException(nameof(listener));

            // Count non-null/non-empty auth properties
            int authCount = 0;
            if (!string.IsNullOrEmpty(options.ApiKey)) authCount++;
            if (!string.IsNullOrEmpty(options.BearerToken)) authCount++;
            if (!string.IsNullOrEmpty(options.SdkToken)) authCount++;

            // Validate exactly-one-auth
            if (authCount == 0)
                throw new ArgumentException("Provide exactly one of: ApiKey, BearerToken, SdkToken", nameof(options));
            if (authCount > 1)
                throw new ArgumentException("Provide exactly one of: ApiKey, BearerToken, SdkToken", nameof(options));

            // Create adapter
            var adapter = new WebSocketListenerAdapter(listener);

            // Convert endpoint
            var uniffiEndpoint = options.Endpoint switch
            {
                WebSocketEndpoint.Stock => uniffi.marketdata_uniffi.WebSocketEndpoint.Stock,
                WebSocketEndpoint.FutOpt => uniffi.marketdata_uniffi.WebSocketEndpoint.FutOpt,
                _ => throw new ArgumentOutOfRangeException(nameof(options.Endpoint))
            };

            // TODO: Current UniFFI WebSocketClient constructors only support API key authentication
            // BearerToken and SdkToken support will be added when UniFFI layer is updated
            try
            {
                if (!string.IsNullOrEmpty(options.ApiKey))
                {
                    // Convert config options to UniFFI record types
                    uniffi.marketdata_uniffi.ReconnectConfigRecord? reconnectRecord = null;
                    if (options.Reconnect != null)
                    {
                        reconnectRecord = new uniffi.marketdata_uniffi.ReconnectConfigRecord(
                            maxAttempts: options.Reconnect.MaxAttempts ?? 0,
                            initialDelayMs: options.Reconnect.InitialDelayMs ?? 0,
                            maxDelayMs: options.Reconnect.MaxDelayMs ?? 0
                        );
                    }

                    uniffi.marketdata_uniffi.HealthCheckConfigRecord? healthCheckRecord = null;
                    if (options.HealthCheck != null)
                    {
                        healthCheckRecord = new uniffi.marketdata_uniffi.HealthCheckConfigRecord(
                            enabled: options.HealthCheck.Enabled ?? false,
                            intervalMs: options.HealthCheck.IntervalMs ?? 0,
                            maxMissedPongs: options.HealthCheck.MaxMissedPongs ?? 0
                        );
                    }

                    _inner = uniffi.marketdata_uniffi.WebSocketClient.NewWithConfig(
                        options.ApiKey,
                        adapter,
                        uniffiEndpoint,
                        reconnectRecord,
                        healthCheckRecord
                    );
                }
                else
                {
                    // For now, only ApiKey is supported in UniFFI WebSocketClient constructors
                    // BearerToken and SdkToken will require UniFFI layer updates
                    throw new NotSupportedException(
                        "WebSocketClient currently only supports ApiKey authentication. " +
                        "BearerToken and SdkToken support will be added in a future update."
                    );
                }

                _reconnectOptions = options.Reconnect;
                _healthCheckOptions = options.HealthCheck;

                // TODO: Apply BaseUrl when UniFFI WebSocketClient exposes base_url() setter
                if (!string.IsNullOrEmpty(options.BaseUrl))
                {
                    // BaseUrl configuration will be implemented when core library supports it
                }
            }
            catch (uniffi.marketdata_uniffi.MarketDataException ex)
            {
                throw new InvalidOperationException($"Failed to create WebSocket client: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Connect to the WebSocket server.
        /// </summary>
        /// <returns>Task that completes when connection is established</returns>
        public Task ConnectAsync() => _inner.Connect();

        /// <summary>
        /// Disconnect from the WebSocket server.
        /// </summary>
        /// <returns>Task that completes when disconnected</returns>
        public Task DisconnectAsync() => _inner.Disconnect();

        /// <summary>
        /// Subscribe to a market data channel for a symbol.
        /// </summary>
        /// <param name="channel">Channel name: "trades", "candles", "books", "meta"</param>
        /// <param name="symbol">Symbol to subscribe (e.g., "2330" for TSMC)</param>
        /// <returns>Task that completes when subscription is confirmed</returns>
        public Task SubscribeAsync(string channel, string symbol) => _inner.Subscribe(channel, symbol);

        /// <summary>
        /// Unsubscribe from a market data channel for a symbol.
        /// </summary>
        /// <param name="channel">Channel name</param>
        /// <param name="symbol">Symbol to unsubscribe</param>
        /// <returns>Task that completes when unsubscription is confirmed</returns>
        public Task UnsubscribeAsync(string channel, string symbol) => _inner.Unsubscribe(channel, symbol);

        /// <summary>
        /// Whether the client is currently connected to the server.
        /// </summary>
        public bool IsConnected => _inner.IsConnected();

        /// <summary>
        /// Whether the client has been shut down.
        /// </summary>
        public bool IsClosed => _inner.IsClosed();

        /// <summary>
        /// Send a ping message to the server.
        /// </summary>
        /// <param name="state">Optional state string echoed back in the pong response</param>
        /// <returns>Task that completes when the ping is sent</returns>
        public Task PingAsync(string? state = null) => _inner.Ping(state);

        /// <summary>
        /// Query the server for current subscriptions.
        /// The response arrives via the OnMessage callback.
        /// </summary>
        /// <returns>Task that completes when the query is sent</returns>
        public Task QuerySubscriptionsAsync() => _inner.QuerySubscriptions();

        private void ThrowIfDisposed()
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(WebSocketClient));
        }

        /// <inheritdoc/>
        public void Dispose()
        {
            if (_disposed) return;
            _inner?.Dispose();
            _disposed = true;
        }
    }
}
